pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1158 @@
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	if slice.ContainsString(clusterCIDR.GetFinalizers(), clusterCIDRFinalizer, nil) {
+		if err := r.deleteClusterCIDR(clusterCIDR); err != nil {
+			return err
+		}
+		// Remove the finalizer as delete is successful.
+		cccCopy := clusterCIDR.DeepCopy()
+		cccCopy.ObjectMeta.Finalizers = slice.RemoveString(cccCopy.ObjectMeta.Finalizers, clusterCIDRFinalizer, nil)
+		if _, err := r.client.NetworkingV1alpha1().ClusterCIDRs().Update(context.TODO(), clusterCIDR, metav1.UpdateOptions{}); err != nil {
+			klog.V(2).Infof("Error removing finalizer for ClusterCIDR %s: %v", clusterCIDR.Name, err)
+			return err
+		}
+		klog.V(2).Infof("Removed finalizer for ClusterCIDR %s", clusterCIDR.Name)
+	}
+	return nil
+}
+
+// deleteClusterCIDR Deletes and unmaps the ClusterCIDRs from the cidrMap.
+func (r *multiCIDRRangeAllocator) deleteClusterCIDR(clusterCIDR *networkingv1alpha1.ClusterCIDR) error {
+
+	labelSelector, err := r.nodeSelectorKey(clusterCIDR)
+	if err != nil {
+		return fmt.Errorf("unable to delete cidr: %w", err)
+	}
+
+	clusterCIDRSetList, ok := r.cidrMap[labelSelector]
+	if !ok {
+		klog.Infof("Label %s not found in CIDRMap, proceeding with delete", labelSelector)
+		return nil
+	}
+
+	for i, clusterCIDRSet := range clusterCIDRSetList {
+		if clusterCIDRSet.Name != clusterCIDR.Name {
+			continue
+		}
+
+		// Mark clusterCIDRSet as terminating.
+		clusterCIDRSet.Terminating = true
+
+		// Allow deletion only if no nodes are associated with the ClusterCIDR.
+		if len(clusterCIDRSet.AssociatedNodes) > 0 {
+			return fmt.Errorf("ClusterCIDRSet %s marked as terminating, won't be deleted until all associated nodes are deleted", clusterCIDR.Name)
+		}
+
+		// Remove the label from the map if this was the only clusterCIDR associated
+		// with it.
+		if len(clusterCIDRSetList) == 1 {
+			delete(r.cidrMap, labelSelector)
+			return nil
+		}
+
+		clusterCIDRSetList = append(clusterCIDRSetList[:i], clusterCIDRSetList[i+1:]...)
+		r.cidrMap[labelSelector] = clusterCIDRSetList
+		return nil
+	}
+	klog.V(2).Info("clusterCIDR not found, proceeding with delete", "Name", clusterCIDR.Name, "label", labelSelector)
+	return nil
+}
+
+func (r *multiCIDRRangeAllocator) nodeSelectorKey(clusterCIDR *networkingv1alpha1.ClusterCIDR) (string, error) {
+	var nodeSelector []byte
+	var err error
+
+	if clusterCIDR.Spec.NodeSelector != nil {
+		nodeSelector, err = clusterCIDR.Spec.NodeSelector.Marshal()
+	} else {
+		nodeSelector, err = defaultNodeSelector()
+	}
+
+	if err != nil {
+		return "", err
+	}
+
+	return string(nodeSelector), nil
+}
+
+func listClusterCIDRs(kubeClient clientset.Interface) (*networkingv1alpha1.ClusterCIDRList, error) {
+	var clusterCIDRList *networkingv1alpha1.ClusterCIDRList
+	// We must poll because apiserver might not be up. This error causes
+	// controller manager to restart.
+	startTimestamp := time.Now()
+
+	// start with 2s, multiply the duration by 1.6 each step, 11 steps = 9.7 minutes
```

Comment:
is this blocking the controller-manager bootstrap or it runs in parallel?
blocking 10 mins the controller-manager sounds excessive to me, but I don't know how others controller are doing to be honest, I don't know if this is normal or not

Reply:
This is similar to the current behavior in the cidr_allocator for listing nodes: https://github.com/kubernetes/kubernetes/blob/bd9444c1cfd438690dc21c31404dd79ab60dc6cb/pkg/controller/nodeipam/ipam/cidr_allocator.go#L130

https://github.com/kubernetes/kubernetes/blob/bd9444c1cfd438690dc21c31404dd79ab60dc6cb/pkg/controller/nodeipam/ipam/cidr_allocator.go#L57-L58

Reply:
Ok, so we are as good or as bad as before 😁